### ISSUE TITLE
Implementation suggestion

### ISSUE THREAD

**Issue by alekso56** [Bobcat00/PlotMarkers#1]
Hey there, I'm interested in this. However, could you please:
1) Utilize a [shapemarker](https://bluemap.bluecolored.de/wiki/customization/Markers.html#shape-markers) instead of a poi marker?

2) Access the PlotSquared API via PlotSquared.get().getPlotAreaManager().forEachPlotArea(plotarea ->{ plotarea.getPlots() });?

[I'm aiming for something similar to this code](https://github.com/Chicken/BlueMap-Towny/blob/9092ddf431e733d34d8316c00394d182c25cc4fb/src/main/java/codes/antti/bluemaptowny/BlueMapTowny.java#L279). [using this library](https://github.com/TechnicJelle/BMUtils)

Using a shapemarker would help reduce the number of markers and enhance the visualization of grouped plots.

thanks for reading.

**Comment by Bobcat00** [comment]
What shapes does PlotSquared support? Do they have to be rectangular, or can they be, for example, L or U shaped? I thought they supported arbitrary shapes, like circles or triangles.

**Comment by Bobcat00** [comment]
Looking at this further, apparently you can have a group of base plots connected like this:
```
              XXX
              XXX
               X             XXXXXX
               X             X    X
               XXXXXXXXXXXXXXX    X
  XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
  X                      X
  X
  X
  X
```
And this would have to be created as a single BlueMap shape. But it gets worse. Two adjacent base plots are not necessarily connected. They could be connected, or there could be a road between them. And the shape would have to account for it. For that matter, the road could be a hole inside the shape.

I really have no idea how to do this. I could look at the PlotSquared code, but they're approaching it from a different direction - If a person is in a plot, not how to recreate the shape of the plot.

**Comment by alekso56** [comment]
> Looking at this further, apparently you can have a group of base plots connected like this:
> 
> ```
>               XXX
>               XXX
>                X             XXXXXX
>                X             X    X
>                XXXXXXXXXXXXXXX    X
>   XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
>   X                      X
>   X
>   X
>   X
> ```
> 
> And this would have to be created as a single BlueMap shape. But it gets worse. Two adjacent base plots are not necessarily connected. They could be connected, or there could be a road between them. And the shape would have to account for it. For that matter, the road could be a hole inside the shape.
> 
> I really have no idea how to do this. I could look at the PlotSquared code, but they're approaching it from a different direction - If a person is in a plot, not how to recreate the shape of the plot.

I would assume it would turn out something like this, (just a rough sketch will probably not work)
hybrid is the newest generator, classic is the older one, square is the base for both.
so it goes hybrid inherits classic, classic inherits square.
some people more intelligent than me have made a library for making cheese from cells of any size, if offsets are not used, this should work fine.  https://github.com/TechnicJelle/BMUtils
```java		
PlotSquared.get().getPlotAreaManager().forEachPlotArea(plotarea ->{
			    
			    if(plotarea instanceof HybridPlotWorld) {
				HybridPlotWorld newest = (HybridPlotWorld) plotarea;
				
			    }else if(plotarea instanceof ClassicPlotWorld ) {
				ClassicPlotWorld classic = (ClassicPlotWorld) plotarea;
				
			    }else if(plotarea instanceof SquarePlotWorld) {
				SquarePlotWorld base = (SquarePlotWorld) plotarea;
				
				base.forEachBasePlot(coreOfMergedPlots -> {
				    Set<Plot> everythingInMergedPlots = coreOfMergedPlots.getConnectedPlots();
				    everythingInMergedPlots.add(coreOfMergedPlots);
				    
				    Vector2i[] chunks = everythingInMergedPlots.stream().map((plot) -> new Vector2i(plot.getId().getX(), plot.getId().getY())).toArray(Vector2i[]::new);
				    
			            Cheese cheese = Cheese.createSingleFromCells(new Vector2d(base.PLOT_WIDTH+base.ROAD_WIDTH,base.PLOT_WIDTH+base.ROAD_WIDTH), chunks);
			            
			            ShapeMarker plotMarker = new ShapeMarker.Builder()
		                                .label(coreOfMergedPlots.getId().toCommaSeparatedString())
		                                .detail(coreOfMergedPlots.getAlias())
		                                .lineColor(new Color("#" + hex + opacity))
		                                .lineWidth(3)
		                                .fillColor(new Color("#" + hex + opacity))
		                                .depthTestEnabled(false)
		                                .shape(cheese.getShape(), (float) base.getMinBuildHeight())
		                                .holes(cheese.getHoles().toArray(Shape[]::new))
		                                .centerPosition()
		                                .build();
			            markerSet.put("plots." + coreOfMergedPlots.getId() + ".area", plotMarker);
			            createMarker(coreOfMergedPlots,markerSet);
				});
			    }
			});
```

**Comment by Bobcat00** [comment]
> (base.PLOT_WIDTH+base.ROAD_WIDTH, base.PLOT_WIDTH+base.ROAD_WIDTH)

This looks like the shape area will cover half the roads on all sides of the plot, which is less than ideal.

**Comment by Bobcat00** [comment]
I actually started working on this. I might have something in January.

**Comment by Bobcat00** [comment]
Give this a try. It adds shape markers to the map. Both the new and old markers can be shown/hidden with browser's map menu.

Let me know what you think. If you don't like it, you can always go back. The blue color is hardcoded; I might add that to the config file. (There are no config changes with this release.)
https://github.com/Bobcat00/PlotMarkers/releases/tag/v2.00-pre1

**Comment by alekso56** [comment]
I tested it quickly on windows, but it failed to copy the icon with an ioException. ill test on a linux install later this week.

**Comment by Bobcat00** [comment]
> I tested it quickly on windows, but it failed to copy the icon with an ioException. ill test on a linux install later this week.

Were you able to try the test version?

**Comment by alekso56** [comment]
Yes, it works perfectly, seems like all the names in bluemap are lowercase, so that caused shapemarkers to not load, i edited the code a bit to make it work on windows also.
![rTvHa7N](https://github.com/user-attachments/assets/24a339a8-f63c-4f9b-8207-16ee53254369)

#4

**Comment by Bobcat00** [comment]
That `if (icon != "")` is pretty suspicious. Oh boy.

The toLowerCase, is that to handle the case where the actual world name has upper-case characters? 

EDIT: I think this has to match the name of the .conf file in the maps  directory. Do you maybe have some sort of mismatch? I'm not sure this is the correct solution, because the POI markers use the same code.

**Comment by alekso56** [comment]
actual world name has uppercase chars, but bluemap always has lowercase chars, unsure if bluemap intentionally does this. i did not test poi markers, i assume they also dont work, i could try changing the config of this plugin but i dont think this plugin is doing anything wrong

**Comment by Bobcat00** [comment]
What are the names of the .conf files in your plugins/BlueMap/maps directory?

We might also be getting confused with world names vs map names.

**Comment by alekso56** [comment]
you are correct, the files inside the bluemap config was set to the correct capitalization, but the filenames were lowercase, i changed them to uppercase and they both work.
i see you have already copied my fixes so ill close the pullrequest.

**Comment by Bobcat00** [comment]
We're not quite done yet. The line where you added the lower case change was just a bad test. Plus there's a problem copying the icon if the map names don't match the world names. So I'm going to work on fixing these, too.

**Comment by Bobcat00** [comment]
Hopefully this will take care of the problems.

https://github.com/Bobcat00/PlotMarkers/releases/tag/v2.00-pre2

**Comment by alekso56** [comment]
Yes, this fixed everything even when the previously broken capitalization was used.

**Comment by Bobcat00** [comment]
Closed by commit 13f2093
